Avoid executing side-effectful graph_module as validation step

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107657
* #107656
* #107643
* __->__ #107271

Dynamo currently runs the real graph module with real inputs as a way to match the return result of graph module with the eager return type. This is unsafe when graph module is side effectful. In the long term, we will get rid of this step. But in the short term, we just fakify the graph module again and run it. 

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

tugsbayasgalan committed:
Fakify param/buffers

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fakify param/buffers"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1157,13 +1157,46 @@ def dynamo_normalization_capturing_compiler(
+                        out = torch.fx.Interpreter(graph).run(*args[params_len:], **kwargs)
+                        return out
+
+                fake_mode_new = _guards.detect_fake_mode(graph_inputs) if _guards.detect_fake_mode(graph_inputs) is not None else fake_mode
```
Don't detect fake mode here

ezyang replied:
use a new fake mode

ezyang replied:
and disable the outer fake mode if there is one maybe_disable_fake_tensor_mode

tugsbayasgalan committed:
Update on "Fakify param/buffers"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1163,7 +1163,35 @@ def result_capturing_wrapper(*graph_inputs):
+                named_buffers = dict(graph.named_buffers(remove_duplicate=False))
+
+                ambient_fake_mode = (
+                    _guards.detect_fake_mode(graph_inputs)
```
After reading the comment at L1152, I am not sure if we should use the `graph_inputs` here? cc @ezyang

ezyang replied:
@tugsbayasgalan if it works to always use the dynamo fake mode I guess this should be preferred for uniformity.

It is "ok" to use the user fake mode in the sense that it is equivalent to what we used to do (do the real compute) but I agree it's more uniform not to.

tugsbayasgalan committed:
Update on "Avoid executing side-effectful graph_module as validation step "


Dynamo currently runs the real graph module with real inputs as a way to match the return result of graph module with the eager return type. This is unsafe when graph module is side effectful. In the long term, we will get rid of this step. But in the short term, we just fakify the graph module again and run it. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Avoid executing side-effectful graph_module as validation step "


Dynamo currently runs the real graph module with real inputs as a way to match the return result of graph module with the eager return type. This is unsafe when graph module is side effectful. In the long term, we will get rid of this step. But in the short term, we just fakify the graph module again and run it. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Avoid executing side-effectful graph_module as validation step "


Dynamo currently runs the real graph module with real inputs as a way to match the return result of graph module with the eager return type. This is unsafe when graph module is side effectful. In the long term, we will get rid of this step. But in the short term, we just fakify the graph module again and run it. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Avoid executing side-effectful graph_module as validation step "


Dynamo currently runs the real graph module with real inputs as a way to match the return result of graph module with the eager return type. This is unsafe when graph module is side effectful. In the long term, we will get rid of this step. But in the short term, we just fakify the graph module again and run it. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Avoid executing side-effectful graph_module as validation step "


Dynamo currently runs the real graph module with real inputs as a way to match the return result of graph module with the eager return type. This is unsafe when graph module is side effectful. In the long term, we will get rid of this step. But in the short term, we just fakify the graph module again and run it. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge
